be/src/storage/lake/general_tablet_writer.cpp

```diff
@@ -114,7 +128,12 @@ Status HorizontalGeneralTabletWriter::flush_segment_writer(SegmentPB* segment) {
         RETURN_IF_ERROR(_seg_writer->finalize(&segment_size, &index_size, &footer_position));
         const std::string& segment_path = _seg_writer->segment_path();
         std::string segment_name = std::string(basename(segment_path));
-        _files.emplace_back(FileInfo{segment_name, segment_size, _seg_writer->encryption_meta()});
+        auto file_info = FileInfo{segment_name, segment_size, _seg_writer->encryption_meta()};
+        if (_seg_writer->shared_file_offset() >= 0) {
+            // This is a shared data file.
+            file_info.shared_file_offset = _seg_writer->shared_file_offset();
```

Comment:
I think you first get the shared-file offset and then write data into shared-file. Am I correct?

If it is, how do you control concurrency between writers of different tablets? If two tablet writers write simultaneously, will these two tablet writers obtain the same offset here?

Reply:
You can check the function inside shared_file:
```
StatusOr<int64_t> SharedWritableFileContext::appendv(const std::vector<Slice>& slices, const FileEncryptionInfo& info) {
    // Use lock to make sure each thread will write to the shared file in order.
    std::lock_guard<std::mutex> l(_shared_file_mutex);
    // Get the offset of current file in the shared file.
    int64_t shared_file_offset = _shared_file->size();
    // Append the slices to the shared file.
    _shared_file->set_encryption_info(info);
    RETURN_IF_ERROR(_shared_file->appendv(slices.data(), slices.size()));
    return shared_file_offset;
}
```
All tablet writer share same `SharedWritableFileContext`, and there is a mutex `_shared_file_mutex` to make sure they could only append one by one